src/Compilers/CSharp/Portable/Errors/DiagnosticInfoWithSymbols.cs

```diff
@@ -20,5 +21,17 @@ internal DiagnosticInfoWithSymbols(bool isWarningAsError, ErrorCode errorCode, o
         {
             this.Symbols = symbols;
         }
+
+        public override bool Equals(object obj)
+        {
+            return obj is DiagnosticInfoWithSymbols diws && 
+                   this.Symbols.Equals(diws.Symbols) && 
+                   base.Equals(obj);
+        }
```

Comment:
Ah, thanks for pointing out the `GetType()` comparison in `DiagnosticInfo.Equals`.

Regarding `abstract` and `sealed`: If all `DiagnosticInfo` types are either `abstract` or `sealed`, and if the `abstract` classes provide `abstract` overrides of `Equals` and `GetHashCode`, then each `sealed` class would be forced to define those methods (presumably in terms of the derived state, and perhaps using helper methods from the `abstract` classes). And for `Equals`, the first check (`obj is DerivedDiagnosticInfo`) would be sufficient for type checking.